### PR TITLE
#102 - Validate query from frontend against json schema

### LINKED ISSUE #102: Validate query from frontend against json schema
**Describe your problem.**

Currently, there is no validation of the structured query sent by the frontend to the backend, leading to possibly malformed (or in the worst case fraudulent) queries being transferred.

**Describe the solution you'd like**

Check queries sent on the "run-query" endpoint against [the sq v2 query schema](https://github.com/num-codex/codex-structured-query/blob/structured-query-v2/structured-query/json-schema/2021_10_18_StructuredQueryV2Schema.json).
If validation fails, return 422 or 400.

Also add a switch to turn validation on or off.

### PR DESCRIPTION
**Pull request by michael-82**
**What's in the PR**
* the [json schema for structured query v2](https://github.com/num-codex/codex-structured-query/blob/structured-query-v2/structured-query/json-schema/2021_10_18_StructuredQueryV2Schema.json) was added and is used to validate incoming queries on the /run-query endpoint against it
* if validation fails, a 422 error is thrown - otherwise nothing changes
* **breaking change**: the run-query method was returning strange responses, always with status code 200 ok and the real code somewhere in the response json body. this was changed to a regular response. However, this currently breaks compatibility with the frontend. @thkoehler11 was involved in the decision to still change it and then adapt the frontend to it.

**How to test manually**
* there is a test class that checks if several schema violations are correctly detected
* otherwise, use a rest client of your choice and check if valid and invalid queries are correctly handled.

**Examples**
Please note that the url may be different on your setup

Call with valid query:
```sh
curl --request POST \
  --url http://localhost:8090/api/v1/query-handler/run-query \
  --header 'Content-Type: application/json' \
  --data '{
  "version": "http://to_be_decided.com/draft-2/schema#",
  "display": "Beispiel-Query",
  "inclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "F",
              "display": "female",
              "system": "https://fhir.loinc.org/CodeSystem/$lookup?system=http://loinc.org&code=LL2191-6",
              "version": ""
            },
            {
              "code": "M",
              "display": "male",
              "system": "https://fhir.loinc.org/CodeSystem/$lookup?system=http://loinc.org&code=LL2191-6",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "a",
            "display": "Jahr"
          },
          "value": 18
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00",
            "display": "F00",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "F09",
            "display": "F09",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ],
  "exclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "male",
              "display": "male",
              "system": "",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "year",
            "display": "Jahr"
          },
          "value": 65
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00.9",
            "display": "F00.9",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "8310-5",
            "display": "Körpertemperatur",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-range",
          "unit": {
            "code": "Cel",
            "display": "°C"
          },
          "minValue": 35,
          "maxValue": 39
        },
        "attributeFilters": [
          {
            "attributeCode": {
              "code": "method",
              "system": "abide",
              "display": "method"
            },
            "type": "concept",
            "selectedConcepts": [
              {
                "code": "LA9370-3",
                "system": "http://loinc.org",
                "display": "Axillary"
              }
            ]
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ]
}
'
```
Call with invalid query (no inclusion criteria present):
```sh
curl --request POST \
  --url http://localhost:8090/api/v1/query-handler/run-query \
  --header 'Content-Type: application/json' \
  --data '{
  "version": "http://to_be_decided.com/draft-2/schema#",
  "display": "Beispiel-Query",
  "exclusionCriteria": [
    [
      {
        "termCodes": [
          {
            "code": "LL2191-6",
            "display": "Geschlecht",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "concept",
          "selectedConcepts": [
            {
              "code": "male",
              "display": "male",
              "system": "",
              "version": ""
            }
          ]
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "30525-0",
            "display": "Alter",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-comparator",
          "comparator": "gt",
          "unit": {
            "code": "year",
            "display": "Jahr"
          },
          "value": 65
        }
      }
    ],
    [
      {
        "termCodes": [
          {
            "code": "F00.9",
            "display": "F00.9",
            "system": "http://fhir.de/CodeSystem/dimdi/icd-10-gm"
          }
        ]
      },
      {
        "termCodes": [
          {
            "code": "8310-5",
            "display": "Körpertemperatur",
            "system": "http://loinc.org"
          }
        ],
        "valueFilter": {
          "type": "quantity-range",
          "unit": {
            "code": "Cel",
            "display": "°C"
          },
          "minValue": 35,
          "maxValue": 39
        },
        "attributeFilters": [
          {
            "attributeCode": {
              "code": "method",
              "system": "abide",
              "display": "method"
            },
            "type": "concept",
            "selectedConcepts": [
              {
                "code": "LA9370-3",
                "system": "http://loinc.org",
                "display": "Axillary"
              }
            ]
          }
        ],
        "timeRestriction": {
          "beforeDate": "2021-10-09",
          "afterDate": "2021-09-09"
        }
      }
    ]
  ]
}
'
```

### CODE REVIEW

**Review by DiCanio — CHANGES_REQUESTED** [review]
Maybe we can even shift to a more "Spring" approach by using the following dependency:
https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-validation

With this most of the implementation logic can be kept. Making use of it is simply a matter of adding a `@Valid` annotation to the appropriate endpoint field and implementing a constraint validator (simply re-use the implementation logic used in `QueryValidator`). Doing so even takes care of translating any error to the correct status codes (and response in general).

File: src/main/java/de/numcodex/feasibility_gui_backend/query/QueryHandlerRestController.java [hunk 1]
```diff
@@ -26,34 +30,56 @@
...
     this.apiBaseUrl = apiBaseUrl;
+    this.queryValidator = queryValidator;
   }
 
   @PostMapping("run-query")
-  public Response runQuery(
+  public ResponseEntity<Object> runQuery(
       @RequestBody StructuredQuery query, @Context HttpServletRequest httpServletRequest) {
 
+    if (queryValidationEnabled) {
```
[the remark this answers is not in the document]
Comment (michael-82) [reply src/main/java/de/numcodex/feasibility_gui_backend/query/QueryHandlerRestController.java]: Not really sure if I exactly got what you meant. However I moved the "enabled" check to the validator itself

File: src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  public void validate(StructuredQuery structuredQuery)
+      throws ValidationException, FileNotFoundException, JSONException {
+    try (InputStream inputStream = QueryValidator.class.getResourceAsStream(JSON_SCHEMA)) {
+      var jsonSchema = new JSONObject(new JSONTokener(inputStream));
+      var jsonSubject = new JSONObject(jsonUtil.writeValueAsString(structuredQuery));
+      SchemaLoader loader = SchemaLoader.builder()
+          .schemaJson(jsonSchema)
+          .draftV7Support()
+          .build();
+      var schema = loader.load().build();
```
[the remark this answers is not in the document]
Comment (michael-82) [reply src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java]: moved it there

File: src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidatorSpringConfig.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class QueryValidatorSpringConfig {
+
+    @Bean
+    public QueryValidator createQueryValidator(ObjectMapper objectMapper) {
```
[the remark this answers is not in the document]
Comment (michael-82) [reply src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidatorSpringConfig.java]: I changed it to create a new one

File: src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.everit.json.schema.ValidationException;
+import org.everit.json.schema.loader.SchemaLoader;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+@Slf4j
```
[the remark this answers is not in the document]
Comment (michael-82) [reply src/main/java/de/numcodex/feasibility_gui_backend/query/validation/QueryValidator.java]: my bad - it is used now

**Comment by michael-82** [comment]
> Maybe we can even shift to a more "Spring" approach by using the following dependency: https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-validation
> 
> With this most of the implementation logic can be kept. Making use of it is simply a matter of adding a `@Valid` annotation to the appropriate endpoint field and implementing a constraint validator (simply re-use the implementation logic used in `QueryValidator`). Doing so even takes care of translating any error to the correct status codes (and response in general).

I changed it accordingly. Not 100% sure if this is the correct/best way to do it though.

**Comment by DiCanio** [comment]
I made some changes. Feel free to check whether you agree with them.

- rename and move classes to the API package (clarifies use case)
- switch to startup configuration (no runtime checks within the validator)
- add tests on a REST controller level (they are important)
- revise some exception handling passages

Furthermore, we do not have a test case on a REST controller level for when validation is turned off. I'm still not happy with that unfounded decision of being able to turn validation off. At this point I repeatedly asked for the reasoning behind this without ever getting a technical answer. Thus, please simply go ahead and take care of any potential fallout.

**Comment by DiCanio** [comment]
Just noticed, that we also need an entry within the README documenting the validation switch.

Other than that, please see the comment above.

**Pull request merged**